[StructuralMechanicsApplication][ShapeOpt] Added von mises stress sensitivities

dbaumgaertner opened the pull request:
This PR:

1) adds a new stress type to the existing adjoint local stress sensitivity analysis (only for solids)
2) adds a new response function (adjoint_max_stress_response), which performs a sensitivity analysis for the max stress within a specified region (simple agglomeration technique)
3) makes the two new features accessible to the ShapeOptimizationApplication and adds a corresponding tests example. Some impressions using the latter are given below:
4) Adds a test to do sensitivity verification against finite differencing, respectively sensitivity verification for different step sizes in a semianalytic approach. The corresponding run-scripts are generic an may be easily adjusted to other scenarios. 

![local_stress_opt_hook_cropped](https://user-images.githubusercontent.com/25897162/53947024-3c8b2e00-40c5-11e9-8f36-e859728118bf.gif)
Figure 1: Loaded hook where the mean von Mises stress of a particular element is minimized (the element is located where the bump forms out and the stresses drop significantly). This optimization uses feature 1) from the list above.

![max_stress_opt_hook_cropped](https://user-images.githubusercontent.com/25897162/53947463-1619c280-40c6-11e9-8fda-35789fa089a8.gif)
Figure 2: Loaded hook where the maximum von Mises stress in the part is minimized. Note, the location of the max value changes. This utilizes feature 2) in the list above. Note also, the region where the max value is evaluated may be customized. So it is possible to take any sub-model part like e.g. only the bend or the support area.  

There are still unit tests missing for the sensitivty analysis
- [ ] Create unit tests using cpp suite  in StructuralMechanicsApp
- [x] Create unit tests for response in StructuralMechanicsApp
- [x] #4764 merged
- [x] Create unit test in shape optimization for max stress response
- [x] Include sensitivity verification scripts in StructuralMechanicsApp

Should not be merged before #4764

armingeiser committed:
Adjoint small displacement element and VON_MISES_STRESS

armingeiser committed:
first example

dbaumgaertner committed:
introduce solid elements in adjoint process

dbaumgaertner committed:
added external adjoint response function

dbaumgaertner committed:
remove automatic step size adaption

dbaumgaertner committed:
extended optimization response function list

dbaumgaertner committed:
added test for external response function

dbaumgaertner committed:
remove dependency on hdf5

dbaumgaertner committed:
Changed DesignLoggerGid to new standard behavior of GidOutputProcess

dbaumgaertner committed:
forgetten change

dbaumgaertner committed:
Merge branch 'master' into feature-lifetime-opt-interface

dbaumgaertner committed:
added test example to outline process flow

dbaumgaertner committed:
Merge branch 'master' into StructMech/solid-stress-opt

dbaumgaertner committed:
Merge branch 'feature-lifetime-opt-interface' into StructMech/testing-solid-stress-opt

dbaumgaertner committed:
added verification test example

dbaumgaertner committed:
remove hdf process and update processes to new format

dbaumgaertner committed:
turn test problem into unconstrained one

dbaumgaertner committed:
introducing variables to output dFdU and dFdX

dbaumgaertner committed:
Merge branch 'master' into StructMech/testing-solid-stress-opt

dbaumgaertner committed:
Merge branch 'master' into StructMech/testing-solid-stress-opt

dbaumgaertner committed:
removing wrong merge

dbaumgaertner committed:
minor cleanup

dbaumgaertner committed:
removed unwanted leftover from merged branch

dbaumgaertner committed:
corrections of wrong commit

dbaumgaertner committed:
minor

dbaumgaertner committed:
minor

dbaumgaertner committed:
remove problems with adjoint settings

1) failing adjoint model part import if primal uses "input_model_part"
2) primal result is not copied when using separate adjoint settings

dbaumgaertner committed:
customized adjoint small disp elements to cover von mises

dbaumgaertner committed:
added max stress response

agglomerates values over a specified domain and uses its max value

dbaumgaertner committed:
added max stress constraint to shape opt

dbaumgaertner committed:
added test example to shape optimization

dbaumgaertner committed:
minor

dbaumgaertner committed:
restore adaptive pertubation

dbaumgaertner committed:
minor

armingeiser commented on applications/ShapeOptimizationApplication/test_examples/09_VonMises_stress_minimization_hook/adjoint_parameters.json:
```diff
@@ -0,0 +1,125 @@
+{
```
Do the automatic settings work for your case?

philbucher commented on applications/ShapeOptimizationApplication/test_examples/09_VonMises_stress_minimization_hook/run_optimization.py:
```diff
@@ -0,0 +1,18 @@
+# Import Kratos core and apps
+from KratosMultiphysics import *
+from KratosMultiphysics.ShapeOptimizationApplication import *
+import optimizer_factory # from ShapeOptimizationApplication
```
```suggestion
from KratosMultiphysics.ShapeOptimizationApplication import optimizer_factory
```

philbucher commented on applications/ShapeOptimizationApplication/test_examples/09_VonMises_stress_minimization_hook/run_optimization.py:
```diff
@@ -0,0 +1,18 @@
+
+# Import Kratos core and apps
+from KratosMultiphysics import *
+from KratosMultiphysics.ShapeOptimizationApplication import *
```
```suggestion
import KratosMultiphysics.ShapeOptimizationApplication as KSO
```

philbucher commented on applications/ShapeOptimizationApplication/test_examples/09_VonMises_stress_minimization_hook/run_optimization.py:
```diff
@@ -0,0 +1,18 @@
+from __future__ import print_function, absolute_import, division
+
+# Import Kratos core and apps
+from KratosMultiphysics import *
```
```suggestion
import KratosMultiphysics as KM
```

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_small_displacement_element.cpp:
```diff
@@ -0,0 +1,201 @@
+        radicant += PK2_k(0,0)*PK2_k(0,0) + PK2_k(1,1)*PK2_k(1,1) + PK2_k(2,2)*PK2_k(2,2);
+        radicant -= PK2_k(0,0)*PK2_k(1,1) + PK2_k(0,0)*PK2_k(2,2) + PK2_k(1,1)*PK2_k(2,2);
+        radicant += 3*PK2_k(0,1)*PK2_k(0,1) + 3*PK2_k(0,2)*PK2_k(0,2) + 3*PK2_k(1,2)*PK2_k(1,2);
+        sensitivity_prefactors[k] = 0.5/sqrt(radicant);
```
```suggestion
        sensitivity_prefactors[k] = 0.5/std::sqrt(radicant);
```

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_small_displacement_element.cpp:
```diff
@@ -0,0 +1,201 @@
+
+    for(IndexType k = 0; k < number_integration_points; ++k)
+    {
+        Matrix & PK2_k = stress_tensor[k];
```
```suggestion
        Matrix& r_pk2_stress_k = stress_tensor[k];
```

philbucher reviewed (changes requested):
Cool example!

Blocking for several reasons (only coding, no clue abt the functionalities, this should be checked by others):
- The test-case is too large, please reduce the size, how many MB is this file?
If you want to keep it as is then please put it to the examples-repo
The Kratos-repo is large enough already, it should not grow boundless
- please add a test for the new functionalities
- Please use the correct pythonic import-style as I pointed out

General question, is this element in line with the one of @msandre ?

MFusseder commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_max_stress_response_function.cpp:
```diff
@@ -0,0 +1,247 @@
+                mean_stress += element_stress[i];
+            mean_stress /= stress_vec_size;
+
+            if(mean_stress > max_mean_stress)
```
Is this definition of max stress reasonably? Let's imagine the following stresses values of a domain: (-10, 1, 1.5, 2). Would the better choice not to optimize for the element with stress value -10 instead of the element with 2?

dbaumgaertner replied:
But that would not be a max() type of response function but a max(abs()), right? This is just a different way of "combining" / "agglomerating" the stresses, much like the KS-function would be another option. Introducing agglomeration techniques would be very cool. But it is definitely a task for another PR.

armingeiser replied:
For VonMises stresses it is of course not a problem, but i do see Martins point.

MFusseder commented on applications/StructuralMechanicsApplication/python_scripts/structural_response.py:
```diff
@@ -374,11 +374,10 @@ def _SynchronizeAdjointFromPrimal(self):
-            variable_utils = KratosMultiphysics.VariableUtils()
-            for variable in self.primal_state_variables:
-                variable_utils.CopyModelPartNodalVar(variable, self.primal_model_part, self.adjoint_model_part, 0)
+        Logger.PrintInfo("> Transfer primal state to adjoint model part.")
```
Why do you removed the if "auto" check? I case one uses hdf5 the primal solution is written twice to the adjoint model part. Or I am wrong?

dbaumgaertner replied:
You are right. The problem is, that if somebody specifies the adjoint parameters without the "auto" feature (only because the output of a special nodal variable is needed), then one is also forced to specify an HDF5 in- and output process, which has further, significant impacts (compiled HDF5-APP). On the other hand, doing twice the job is also bad. I am very open to suggestions to solve this problem.

msandre commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_processes/replace_elements_and_conditions_for_adjoint_problem_process.cpp:
```diff
@@ -117,6 +118,12 @@ namespace Kratos
 
                     (*it.base()) = p_element;
                 }
+                else if (element_name == "AdjointFiniteDifferenceSmallDisplacementElement")
```
This is outdated. You need to merge master and use the new style to fix your merge conflicts.

msandre commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_small_displacement_element.cpp:
```diff
@@ -0,0 +1,201 @@
+namespace Kratos
+{
+
+AdjointFiniteDifferenceSmallDisplacementElement::AdjointFiniteDifferenceSmallDisplacementElement(Element::Pointer pPrimalElement)
```
Check current master for how to do this consistently with other adjoint elements.

msandre commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_small_displacement_element.cpp:
```diff
@@ -0,0 +1,201 @@
+    {
+        Matrix & PK2_k = stress_tensor[k];
+        double radicant = 0.0;
+        radicant += PK2_k(0,0)*PK2_k(0,0) + PK2_k(1,1)*PK2_k(1,1) + PK2_k(2,2)*PK2_k(2,2);
```
Does this work in 2d?

dbaumgaertner replied:
This will not work for 2D like this. But 2D also requires a bit more thinking and testing. I would want to leave this to a future PR. Instead, a corresponding error is dropped now.

msandre reviewed (dismissed):
Please add a unit test for your element to the structural mechanics application. You can use the setup in `StructuralMechanicsApplication/tests/cpp_tests/test_solid_static_sensitivity.cpp` to quickly create a minimal test. Also check if it works in 2D.

Please also add unit tests for your response functions. They can be tested in isolation using simple finite differences.

@philbucher already commented it, but it's important not to add large examples to the repo, regardless of which application.

dbaumgaertner committed:
Merge branch 'master' into StructMech/testing-solid-stress-opt

dbaumgaertner committed:
adusted analysis to new replacement process

dbaumgaertner committed:
avoid local sensitivity assembly if element has no contribution

dbaumgaertner committed:
introduce missing resize in element and condition base class

philbucher reviewed (changes requested):
the mdpa file you are adding is gigantic, could you please move this to the examples-repo to not further increase the size of the kratos-main repo?

dbaumgaertner commented:
I am about to clean everything.

dbaumgaertner committed:
removed unnecessary big example

dbaumgaertner committed:
added template scripts for sensitivity verification

this exemplary test case can be easily be used and adjusted to test other cases or response functions

dbaumgaertner committed:
minor

dbaumgaertner committed:
added shape opt unit test for max stress objective

dbaumgaertner committed:
added stress opt test to unit test

dbaumgaertner committed:
adjusted import in unit tests to standard python type

dbaumgaertner committed:
added conditions to ignore list in adjoint static solver

MFusseder commented on kratos/utilities/sensitivity_builder.cpp:
```diff
@@ -291,9 +294,12 @@ void CalculateNonHistoricalSensitivities(const SensitivityVariables<TDataType>&
-        AssembleOnDataValueContainer(*rVariables.pOutputVariable,
-                                     builder.LocalSensitivity, rElement.Data());
+        // if rElement does not contribute to local sensitivity, skip assembly
+        if (builder.LocalSensitivity.size() != 0){
```
This will not work, since in the case a element does not contribute to the sensitivity the size of the pseudo-load computed in `CalculateSensitivityMatrix` is not sized to zero. @dbaumgaertner I will contact you in person in order to talk about a concept on how to deal with this topic because its also on my TODO-list.

dbaumgaertner committed:
minor

dbaumgaertner committed:
check for proper dimension

MFusseder commented:
@armingeiser and me created a new branch called StructMech/preparations-for-topology-optimization based on this one. In this new branch we already merged the master and solved conflicts. In order to that this changes can be cherry picked.

philbucher commented:
Can this PR be closed then?

dbaumgaertner commented:
No! It will be updated and merged soon. There are just some other things with priority first.

dbaumgaertner committed:
extended unit test for max stress

dbaumgaertner committed:
Merge branch 'master' into StructMech/testing-solid-stress-opt

dbaumgaertner committed:
corrections after merge

dbaumgaertner committed:
forgotten correction

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/stress_response_definitions.cpp:
```diff
@@ -456,5 +459,29 @@ void StressCalculation::CalculateStressOnNodeBeam(Element& rElement,
 }
 
+
+void StressCalculation::CalculateStressOnGPSmallDisplacement(Element& rElement,
```
why is this not a `const` fct?

dbaumgaertner committed:
added response unit test to StructuralMechanicsApp

dbaumgaertner commented:
@armingeiser I added a unit test for the max stress response (no aggregation yet). 

With this, I want to finally bring the stress optimization to the master. This shall prepare the code for further features such as stress aggregation. Could you check the last few commits and see if there is something problematic? Note that I also made some changes in the StructuralMechanicsApp (ping @philbucher).

philbucher commented on applications/StructuralMechanicsApplication/tests/response_function_tests/adjoint_max_stress_response_parameters.json:
```diff
@@ -0,0 +1,30 @@
+        "domain_size"     : 3,
+        "time_step"       : 1.0,
+        "start_time"      : 0.0,
+        "end_time"        : 1.0,
```
I think those are now under solver_settings

philbucher commented on applications/StructuralMechanicsApplication/tests/response_function_tests/adjoint_max_stress_response_parameters.json:
```diff
@@ -0,0 +1,30 @@
+        "gradient_mode"        : "semi_analytic",
+        "step_size"            : 1.0e-6,
+        "critical_part_name"   : "stress_region",
+        "stress_type"          : "VON_MISES_STRESS",
```
style-guide: `von_mises_stress`

armingeiser replied:
This follows thee existing code, see comment below.

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_response.py:
```diff
@@ -399,6 +399,12 @@ def _GetAdjointParameters(self):
                 KratosMultiphysics.Logger.PrintWarning(__name__, depr_msg)
                 solver_settings.RemoveValue("scheme_settings")
 
+            if solver_settings["model_import_settings"]["input_type"].GetString() == "use_input_model_part":
```
this is not related to this PR is it?

armingeiser replied:
It is because it is the first time a response (stress . adjoint) is used in optimization.

dbaumgaertner replied:
as @armingeiser said, this is in fact needed, as the adjoint model part always has to be read in as it gets modified later on. The option "use_input_model_part" is hence not applicable.

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_response.py:
```diff
@@ -409,9 +415,9 @@ def _GetAdjointParameters(self):
+            if adjoint_parameters.Has("output_processes"):
                 Logger.PrintInfo("> Output process is removed for adjoint analysis. To enable it define adjoint_parameters yourself.")
-                adjoint_parameters.RemoveValue("output_configuration")
+                adjoint_parameters.RemoveValue("output_processes")
```
I think this is quite a weak mechanism, note that not all processes that generate output are in output-processes

armingeiser replied:
This automatic setup for the adjoint parameters anyway works only for standard cases. If something more complex comes into play, the adjoint parameters can and should be specified manually.

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/stress_response_definitions.h:
```diff
@@ -127,6 +128,11 @@ class StressCalculation
                             Vector& rOutput,
                             const ProcessInfo& rCurrentProcessInfo);
 
+    static void CalculateStressOnGPSmallDisplacement(Element& rElement,
```
better use a namespace in the future

```suggestion
    static void CalculateStressOnGPSmallDisplacement(const Element& rElement,
```

dbaumgaertner replied:
That does not work due to the internal call of  "CalculateOnIntegrationPoints" from the element.

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/stress_response_definitions.cpp:
```diff
@@ -357,7 +360,7 @@ void StressCalculation::CalculateStressOnGPShell(Element& rElement,
 }
 
-void StressCalculation::StressCalculation::CalculateStressBeam(Element& rElement,
+void StressCalculation::CalculateStressBeam(Element& rElement,
```
```suggestion
void StressCalculation::CalculateStressBeam(const Element& rElement,
```

dbaumgaertner replied:
The const expression is not possible here due to the same reason as above.

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_small_displacement_element.cpp:
```diff
@@ -0,0 +1,205 @@
+        << "AdjointFiniteDifferencingSmallDisplacementElement::CalculateStressDisplacementDerivative: Invalid element dimension! Currently only 3D SmallDisplacementElements are supported!" << std::endl;
+
+    // Build vector of variables containing the DOF-variables of the primal problem
+    std::vector<VariableComponent<VectorComponentAdaptor<array_1d<double, 3>>>> primal_solution_variable_list;
```
does this work?

```suggestion
    std::vector<VariableComponent<VectorComponentAdaptor<array_1d<double, 3>>>> primal_solution_variable_list {DISPLACEMENT_X, DISPLACEMENT_Y, DISPLACEMENT_Z};
```

philbucher reviewed (commented):
some comments

armingeiser commented on applications/ShapeOptimizationApplication/tests/sensitivity_analysis_test/run_semi_analytic_step_size_verification.py:
```diff
@@ -0,0 +1,83 @@
+            line_to_write += '%.6E' % Decimal(str(gradient_of_interest[1])) + ",\t"
+            line_to_write += '%.6E' % Decimal(str(gradient_of_interest[2])) + ",\t"
+            line_to_write += str(time.ctime()) + "\n"
+            open_file.write(line_to_write)
```
This is useful but it is not a test - meaning it will not be tested and it will become outdated quite easily.

armingeiser replied:
Maybe we can make another folder with some useful skripts inside?

dbaumgaertner replied:
I believe those scripts can be really useful in the verification of new response functions. But you are right, the problem is that they are not tested. Actually here it is not about a single verification, but only about whether or not the script runs through and so follows the latest Kratos standards. 

So, I am thinking to add the three scripts to the nightly tests such that it is frequently tested if the process is still running. What do you think? The execution time is only a few seconds.

armingeiser replied:
That is fine for me.

armingeiser commented on applications/ShapeOptimizationApplication/tests/sensitivity_analysis_test/run_sensitivity_verification_in_design_space.py:
```diff
@@ -0,0 +1,173 @@
+            line_to_write += '%.6E' % Decimal(str(fd_gradient_y)) + ",\t"
+            line_to_write += '%.6E' % Decimal(str(fd_gradient_z)) + ",\t"
+            line_to_write += str(time.ctime()) + "\n"
+            open_file.write(line_to_write)
```
same

armingeiser commented on applications/ShapeOptimizationApplication/tests/sensitivity_analysis_test/run_sensitivity_verification_in_geometry_space.py:
```diff
@@ -0,0 +1,120 @@
+            line_to_write += '%.6E' % Decimal(str(fd_gradient_y)) + ",\t"
+            line_to_write += '%.6E' % Decimal(str(fd_gradient_z)) + ",\t"
+            line_to_write += str(time.ctime()) + "\n"
+            open_file.write(line_to_write)
```
what is the difference between this and run_semi_analytic_step_size_verification.py ?

dbaumgaertner replied:
In "run_semi_analytic_step_size_verification.py", basically a step-size study is done to see if for different step-sizes similar gradients are obtained by the semi-analytic approach. Whereas, in the "run_sensitivity_verification_in_geometry_space.py", the entire gradient for selected nodes are computed by finite differencing and varying the grid coordinates (no semi-analytic approach here). Theoretically, the latter has to match the former.

armingeiser commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_max_stress_response_function.cpp:
```diff
@@ -0,0 +1,239 @@
+        // Set default traced element in adjoint model part (first in the list of agglomerated elements).
+        ModelPart& adjoint_agglomeration_part = mrAdjointModelPart.GetSubModelPart(mCriticalPartName);
+        auto first_element = *(adjoint_agglomeration_part.ElementsBegin());
+        mpTracedElementInAdjointPart = mrAdjointModelPart.pGetElement(first_element.Id());
```
this should be nullptr before it is properly assigned in CalculateValue

dbaumgaertner replied:
Done. I also added a corresponding error check.

armingeiser commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/stress_response_definitions.cpp:
```diff
@@ -53,6 +53,7 @@ namespace StressResponseDefinitions
             { "MZY", TracedStressType::MZY },
             { "MZZ", TracedStressType::MZZ },
             { "PK2", TracedStressType::PK2 },
+            { "VON_MISES_STRESS", TracedStressType::VON_MISES_STRESS }
```
@philbucher this is where it is used, if we change this we would need to adapt the others as well - definitely part of another MR if necessary.

armingeiser reviewed (changes requested):
I have a few comments, but in general it looks good! Thanks!

dbaumgaertner committed:
added sensitivity verification process to nightly tests

to test the process not a single sensitivity analysis

dbaumgaertner commented:
@armingeiser @philbucher I included all your suggestions. Thanks! So ready for another review.

@armingeiser I included the sensitivity verification scripts as nightly tests just to test the process.

philbucher commented on applications/ShapeOptimizationApplication/tests/sensitivity_verification_process_test/run_sensitivity_verification_in_design_space.py:
```diff
@@ -170,4 +173,15 @@ def OutputModelPart( output_mdpa, output_filename, nodal_results ):
+    kratos_utilities.DeleteFileIfExisting(results_filename)
+    kratos_utilities.DeleteFileIfExisting("structure.post.bin")
+    kratos_utilities.DeleteFileIfExisting("reference_part.post.bin")
+    kratos_utilities.DeleteFileIfExisting("perturbed_part.post.bin")
```
why do you delete them?
In the regular testing routine I would not write them, since they are not used and take up valueable computation time
In case you want to look at them (e.g. while setting up a test or debugging) then you wouldn't want to delete them :)

dbaumgaertner replied:
That is true. In fact, the purpose of those scripts is to look at the produced output. So once you use it, you would deactivate the deletion of the files. The purpose of the test, however, is to guarantee that the process is running so that the script remains "plug and play".

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_max_stress_response_function.h:
```diff
@@ -169,7 +169,7 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) AdjointMaxStressResponseFunct
     ModelPart& mrAdjointModelPart;
     std::string mCriticalPartName;
-    Element::Pointer mpTracedElementInAdjointPart;
+    Element::Pointer mpTracedElementInAdjointPart {nullptr};
```
didn't know this works

dbaumgaertner replied:
Well now that I review it myself, I prefer the more explicit notation with "= nullptr". I changed it. Could you just quickly reapprove? Thanks!

philbucher replied:
I can approve but @armingeiser and @msandre requested changes ...

dbaumgaertner replied:
Yeah true. Anyways they also requested changes, so it will not be possible to merge before their approval. So, of course, we wait for them. It is just because you already gave an approval before.

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_max_stress_response_function.cpp:
```diff
@@ -160,6 +160,9 @@ namespace Kratos
     {
         KRATOS_TRY;
 
+        KRATOS_ERROR_IF(mpTracedElementInAdjointPart == nullptr)
```
`KRATOS_ERROR_IF_NOT(mpTracedElementInAdjointPart)` would also work

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_small_displacement_element.cpp:
```diff
@@ -48,11 +48,7 @@ void AdjointFiniteDifferencingSmallDisplacementElement<TPrimalElement>::Calculat
-    primal_solution_variable_list.push_back(DISPLACEMENT_X);
-    primal_solution_variable_list.push_back(DISPLACEMENT_Y);
-    primal_solution_variable_list.push_back(DISPLACEMENT_Z);
+    std::vector<VariableComponent<VectorComponentAdaptor<array_1d<double, 3>>>> primal_solution_variable_list {DISPLACEMENT_X, DISPLACEMENT_Y, DISPLACEMENT_Z};
```
nice !

philbucher reviewed (dismissed):
ok for me

dbaumgaertner committed:
minor change for improved reading

philbucher commented:
I approved 3 times now, and there are 2 times changes requested

theoretically you should be able to merge since 3 > 2 but it still seems blocked ...

armingeiser reviewed (approved):
Thanks! @msandre could you also remove your blocker?

loumalouomega commented:
Cool, one question, why don't you upload these cool hook examples to the Examples repository? https://github.com/KratosMultiphysics/Examples

armingeiser commented:
@loumalouomega it is: https://github.com/KratosMultiphysics/Examples/tree/master/shape_optimization/01_Strain_Energy_Minimization_3D_Hook ;-)

philbucher commented:
@msandre since you are no longer developing Kratos :( I will dismiss your review on Monday such that further developments are not blocked :) 
(Also I think your concerns were addressed)

philbucher commented:
@dbaumgaertner can you please check your test? It fails in the nightly build
I would say you have to decrease the tolerance
~~~
======================================================================
FAIL: test_execution (structural_response_function_test_factory.TestAdjointMaxStressResponseFunction)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ubuntu/Kratos/applications/StructuralMechanicsApplication/tests/structural_response_function_test_factory.py", line 136, in test_execution
    self.assertAlmostEqual(self.gradient[5][0], 1787255.3702425747, 10)
AssertionError: 1787255.3699018753 != 1787255.3702425747 within 10 places
----------------------------------------------------------------------
~~~

loumalouomega commented:
I would say to use relative tolerance, the values are almost identical in relative terms

philbucher commented:
I agree
In fact I was thinking abt implementing `assertAlmostEqualRelative` in `KratosUnittest`

philbucher commented:
fixed in #5713
